src/Microsoft.ML.TimeSeries/RootCauseAnalyzer.cs

```diff
@@ -0,0 +1,753 @@
+            GetRootCauseDirectionAndScore(dimPointMapping, src.AnomalyDimension, dst, _beta, pointTree, src.AggType, src.AggSymbol);
+
+            return dst;
+        }
+
+        protected List<Point> GetTotalPointsForAnomalyTimestamp(RootCauseLocalizationInput src)
+        {
+            MetricSlice slice = src.Slices.Single(slice => slice.TimeStamp.Equals(src.AnomalyTimestamp));
+            return slice.Points;
+        }
+
+        protected DimensionInfo SeperateDimension(Dictionary<string, Object> dimensions, Object aggSymbol)
+        {
+            DimensionInfo info = DimensionInfo.CreateDefaultInstance();
+            foreach (KeyValuePair<string, Object> entry in dimensions)
+            {
+                string key = entry.Key;
+                if (aggSymbol.Equals(entry.Value))
+                {
+                    info.AggDims.Add(key);
+                }
+                else
+                {
+                    info.DetailDims.Add(key);
+                }
+            }
+
+            return info;
+        }
+
+        protected Tuple<PointTree, PointTree, Dictionary<string, Point>> GetPointsInfo(RootCauseLocalizationInput src, DimensionInfo dimensionInfo)
+        {
+            PointTree pointTree = new PointTree();
+            PointTree anomalyTree = new PointTree();
+            Dictionary<string, Point> dimPointMapping = new Dictionary<string, Point>();
+
+            List<Point> totalPoints = GetTotalPointsForAnomalyTimestamp(src);
+            Dictionary<string, Object> subDim = GetSubDim(src.AnomalyDimension, dimensionInfo.DetailDims);
+
+            foreach (Point point in totalPoints)
+            {
+                if (ContainsAll(point.Dimension, subDim))
+                {
+                    if (!dimPointMapping.ContainsKey(GetDicCode(point.Dimension)))
+                    {
+                        dimPointMapping.Add(GetDicCode(point.Dimension), point);
+                        bool isValidPoint = point.IsAnomaly == true;
+                        if (ContainsAll(point.Dimension, subDim))
+                        {
+                            BuildTree(pointTree, dimensionInfo.AggDims, point, src.AggSymbol);
+
+                            if (isValidPoint)
+                            {
+                                BuildTree(anomalyTree, dimensionInfo.AggDims, point, src.AggSymbol);
+                            }
+                        }
+                    }
+                }
+            }
+
+            return new Tuple<PointTree, PointTree, Dictionary<string, Point>>(pointTree, anomalyTree, dimPointMapping);
+        }
+
+        protected Dictionary<string, Object> GetSubDim(Dictionary<string, Object> dimension, List<string> keyList)
+        {
+            return new Dictionary<string, object>(keyList.Select(dim => new KeyValuePair<string, object>(dim, dimension[dim])).ToDictionary(kvp => kvp.Key, kvp => kvp.Value));
+
```

Comment:
Do you need the the ToDictionary at the end? Aren't you already returning a new Dictionary?

Reply:
> Do you need the the ToDictionary at the end? Aren't you already returning a new Dictionary?

If I don't add the toDictionary at the end, there will be a compile error, say that "Cannot convert IEnumerable<KeyValuePair<string, object>> to IDictionary"